CC-21664 Migration guide - Upgrade to Webpack v5 (#1476)
spryker/spryker-docs: docs/scos/dev/front-end-development/migration-guide-upgrade-to-webpack-v5.md

---
title: Migration guide - Upgrade to Webpack v5
description: Use the guide to update versions of the Webpack and related modules.
template: module-migration-guide-template
---

> ## Upgrading from version 4.* to version 5.*

Reviewer: ```suggestion
## Upgrading Webpack from version 4.* to version 5.*
```

This document provides instructions for upgrading Webpack to version 5 in your Spryker project.

## Overview

The Webpack v5 was released 2 years ago.
So we need to migrate as well to unblock future upgrades of other dependencies.

*Estimated migration time: 1h*

## 1) Update dependencies

{% info_block infoBox "Note" %}

To ensure that dependencies will install correctly, delete `node_modules` and `package-lock.json` before installation 

> using the next command `rm -rf node_modules && rm -rf package-lock.json`.

Author: Removed section

{% endinfo_block %}

1. Update the following dependencies to the new version:

```json
{
    // update
    "@types/webpack": "~5.28.0", // If using Merchant Portal
    "clean-webpack-plugin": "~4.0.0",
    "compression-webpack-plugin": "~10.0.0",
    "copy-webpack-plugin": "~11.0.0",
    "mini-css-extract-plugin": "~2.7.2",
    "optimize-css-assets-webpack-plugin": "~6.0.1",
    "terser-webpack-plugin": "~5.3.6",
    "webpack": "~5.74.0",

>     "webpack-merge": "~5.8.0",

Reviewer: Should we mention the `fast-glob`, `prettier`, `core-js` updates here or it's needed for the angular update?

Author: It's not needed even for angular, just was updated

    // add

>     "postcss": "~8.4.20"

Reviewer: But I don't see the dependency in the [PR](https://github.com/spryker/suite-nonsplit/pull/7449)

Author: `yves:production` requires postCSS version 8

Reviewer: OK, but why don't we have this in our [package.json](https://github.com/spryker/suite-nonsplit/blob/a3d1e41973c6236a5dd37e9e03b185ddc4bfd251/package.json)?

Author: Seems some other dependency covers it

Reviewer: So I think we should add it to avoid an unexpected issue.

Author: Added

}
```

> 2. Run `npm install` to install dependencies and update `package-lock.json` file.

Author: Rephrased

## 2) Update configuration files

**Do the following changes to update configuration files:**

1. `frontend/configs/development-watch.js`:

```js
const merge = require('webpack-merge');
// must be
const { merge } = require('webpack-merge');
```

2. `frontend/configs/development.js`:

> - Rename the `output.jsonpFunction` property to the `chunkLoadingGlobal`.
> - Rename the `optimization.splitChunks.cacheGroups.vendors` property to the `defaultVendors`.

Reviewer: ```suggestion
- Rename the `jsonpFunction` property to the `chunkLoadingGlobal` of the `output` object.
- Rename the `vendors` property to the `defaultVendors` of the `cacheGroups` object.
```

3. `frontend/configs/production.js`:

```js
const merge = require('webpack-merge');
// must be
const { mergeWithCustomize, customizeObject } = require('webpack-merge');
```

```js
const mergeWithStrategy = merge.smartStrategy({
    plugins: 'prepend',
});
// must be
const mergeWithStrategy = mergeWithCustomize({
    customizeObject: customizeObject({
        plugins: 'prepend',
    })
});
```

```js
plugins: [
    new CompressionPlugin({
        filename: '[path].gz[query]',
    }),

>     new CompressionPlugin({ // or `BrotliPlugin`

Reviewer: ```suggestion
    new CompressionPlugin({
```

Author: We still have BrotliPlugin plugin usage on demoshops, I assume customers have it as well

Reviewer: This is not clear.
Also, I've checked 4 demo shops and they have: 
```
new BrotliPlugin({
                asset: '[path].br[query]',
                test: /\.js$|\.css$|\.svg$|\.html$/,
                threshold: 10240,
                minRatio: 0.8
            })
```
So let's have this in the example.

        filename: '[path].br[query]',
        algorithm: 'brotliCompress',
        test: /\.(js|css|html|svg)$/,
        threshold: 10240,
        minRatio: 0.8,
    }),
]
// must be
plugins: [
    new CompressionPlugin({
        filename: '[path][base].br[query]',
        algorithm: 'brotliCompress',
        test: /\.(js|css|html|svg)$/,
        threshold: 10240,
        minRatio: 0.8,
    }),
]
```

> Remove the `cache` property in the `new TerserPlugin({ ... })` minimize plugin.

Reviewer: ```suggestion
Remove the `cache` property in the `new TerserPlugin({ ... })` minimizer plugin.
```

        return [
            {
                from: staticAssetsPath,
                to: appSettings.paths.publicStatic,
                context: appSettings.context,
            },
        ];
    }
    
    return [];
};
```

```js
const getAssetsConfig = (appSettings) => [
    new CleanWebpackPlugin(
        [
            appSettings.paths.public,
            appSettings.paths.publicStatic,
        ],
        {
            root: appSettings.context,
            verbose: true,
            beforeEmit: true,
        }
    ),

    new CopyWebpackPlugin(getCopyConfig(appSettings), {
        context: appSettings.context,
    }),

    new CopyWebpackPlugin(getCopyStaticConfig(appSettings), {
        context: appSettings.context,
    }),
];
// must be
const getAssetsConfig = (appSettings) => [
    new CleanWebpackPlugin({
        cleanOnceBeforeBuildPatterns: [
            appSettings.paths.public,
            appSettings.paths.publicStatic,
        ],
        verbose: true,
    }),

    new CopyPlugin({
        patterns: getCopyConfig(appSettings),
    }),

    new CopyPlugin({
        patterns: getCopyStaticConfig(appSettings),
    }),
];
```

## 2) Build assets

> **Check correct builds for all layers `Yves`/`Zed`/`MP` including `watch`/`production` modes:**

Reviewer: ```suggestion
**Check correct builds for all layers `Yves`/`Zed`/`MP` (if Merchant Portal is used) including `watch`/`production` modes:**
```

1. `Yves`:

```bash
npm run yves
npm run yves:production
npm run yves:watch
```

2. `Zed`:

```bash
npm run zed
npm run zed:production
npm run zed:watch
```

> 3. `MP` (If using Merchant Portal):

Reviewer: ```suggestion
3. `MP` (if Merchant Portal is used):
```

---
title: Migration guide - Upgrade to Webpack v5
description: Use the guide to update versions of the Webpack and related modules.
last_updated: Jan 12, 2023
template: module-migration-guide-template
---

## Upgrading Webpack from version 4.* to version 5.*

This document provides instructions for upgrading Webpack to version 5 in your Spryker project.

## Overview

The Webpack v5 was released two years ago.
To unblock future upgrades of other dependencies, we need to migrate as well.

*Estimated migration time: 1h*

## 1) Update dependencies

1. In `package.json`, update the following dependencies to the new version:

```json
{
    "clean-webpack-plugin": "~4.0.0",
    "compression-webpack-plugin": "~10.0.0",
    "copy-webpack-plugin": "~11.0.0",
    "mini-css-extract-plugin": "~2.7.2",
    "optimize-css-assets-webpack-plugin": "~6.0.1",
    "postcss": "~8.4.20",
    "terser-webpack-plugin": "~5.3.6",
    "webpack": "~5.74.0",
    "webpack-merge": "~5.8.0"
}
```

> - If Merchant Portal is used:

Author: This is not a command, this is a part of .json file, so I've removed it

```json
{
    "@types/webpack": "~5.28.0"
}
```

2. Update and install package dependencies:

```bash
rm -rf node_modules
npm install
```

>    

Reviewer 2: If this is a verification process (line 52), then I suppose we could add a verification block here
```suggestion

{% info_block warningBox "Verification" %}

```

Ensure that the `package-lock.json` file and the `node_modules` folder have been updated.

> 

Reviewer 2: closed info block
```suggestion

{% endinfo_block %}
```

## 2) Update configuration files

To update configuration files, make the following changes:

> 1. `frontend/configs/development-watch.js`:

Reviewer 2: Can we use a complete sentence here and for other js files listed below? For example:
```suggestion
1. In `frontend/configs/development-watch.js`, add the following code:
```

        return [
            {
                from: staticAssetsPath,
                to: appSettings.paths.publicStatic,
                context: appSettings.context,
            },
        ];
    }
    
    return [];
};
```

```js
const getAssetsConfig = (appSettings) => [
    new CleanWebpackPlugin(
        [
            appSettings.paths.public,
            appSettings.paths.publicStatic,
        ],
        {
            root: appSettings.context,
            verbose: true,
            beforeEmit: true,
        }
    ),

    new CopyWebpackPlugin(getCopyConfig(appSettings), {
        context: appSettings.context,
    }),

    new CopyWebpackPlugin(getCopyStaticConfig(appSettings), {
        context: appSettings.context,
    }),
];
// must be
const getAssetsConfig = (appSettings) => [
    new CleanWebpackPlugin({
        cleanOnceBeforeBuildPatterns: [
            appSettings.paths.public,
            appSettings.paths.publicStatic,
        ],
        verbose: true,
    }),

    new CopyPlugin({
        patterns: getCopyConfig(appSettings),
    }),

    new CopyPlugin({
        patterns: getCopyStaticConfig(appSettings),
    }),
];
```

## 2) Build assets

> Check correct builds for all layers `Yves`/`Zed`/`MP` (if Merchant Portal is used) including `watch`/`production` modes:

Reviewer 2: if you use backslash as a comma or "or", then I suggest this rephrasing:
```suggestion
Check correct builds for all layers `Yves`, `Zed`, and `MP` (if Merchant Portal is used), including `watch` and `production` modes:
```

   - Rename the `jsonpFunction` property to the `chunkLoadingGlobal` of the `output` object.
   - Rename the `vendors` property to the `defaultVendors` of the `cacheGroups` object.

3. `frontend/configs/production.js`:

   ```js
   const merge = require('webpack-merge');
   // must be
   const { mergeWithCustomize, customizeObject } = require('webpack-merge');
   ```

   ```js
   const mergeWithStrategy = merge.smartStrategy({
       plugins: 'prepend',
   });
   // must be
   const mergeWithStrategy = mergeWithCustomize({
       customizeObject: customizeObject({
           plugins: 'prepend',
       })
   });
   ```

   ```js
   plugins: [
       new CompressionPlugin({
           filename: '[path].gz[query]',
       }),
   
       new BrotliPlugin({
           asset: '[path].br[query]',
           test: /\.js$|\.css$|\.svg$|\.html$/,
           threshold: 10240,
           minRatio: 0.8
       })
   ]
   // must be
   plugins: [
       new CompressionPlugin({
           filename: '[path][base].br[query]',
           algorithm: 'brotliCompress',
           test: /\.(js|css|html|svg)$/,
           threshold: 10240,
           minRatio: 0.8,
       }),
   ]
   ```

> Remove the `cache` property in the `new TerserPlugin({ ... })` minimizer plugin.

Reviewer: ```suggestion
   Remove the `cache` property in the `new TerserPlugin({ ... })` minimizer plugin.
```

## Overview

The Webpack v5 was released two years ago.
To unblock future upgrades of other dependencies, we need to migrate as well.

*Estimated migration time: 1h*

## 1) Update dependencies

1. In `package.json`, update the following dependencies to the new version:

```json
{
    "clean-webpack-plugin": "~4.0.0",
    "compression-webpack-plugin": "~10.0.0",
    "copy-webpack-plugin": "~11.0.0",
    "mini-css-extract-plugin": "~2.7.2",
    "optimize-css-assets-webpack-plugin": "~6.0.1",
    "postcss": "~8.4.20",
    "terser-webpack-plugin": "~5.3.6",
    "webpack": "~5.74.0",
    "webpack-merge": "~5.8.0"
}
```

- If Merchant Portal is used, run the following command:

```json
{
    "@types/webpack": "~5.28.0"
}
```

2. Update and install package dependencies:

```bash
rm -rf node_modules
npm install
```

{% info_block warningBox "Verification" %}

Ensure that the `package-lock.json` file and the `node_modules` folder have been updated.

## 2) Update configuration files

To update configuration files, make the following changes:

1. `frontend/configs/development-watch.js`:

```js
const merge = require('webpack-merge');
// must be
const { merge } = require('webpack-merge');
```

> 2. `frontend/configs/development.js`:

Reviewer 2: ```suggestion
2. In `frontend/configs/development.js`, do the following:
```

   - Rename the `jsonpFunction` property to the `chunkLoadingGlobal` of the `output` object.
   - Rename the `vendors` property to the `defaultVendors` of the `cacheGroups` object.

> 3. `frontend/configs/production.js`:

Reviewer 2: ```suggestion
3. In `frontend/configs/production.js`, add the following code:
```

```js
const merge = require('webpack-merge');
// must be
const { mergeWithCustomize, customizeObject } = require('webpack-merge');
```

```js
const mergeWithStrategy = merge.smartStrategy({
    plugins: 'prepend',
});
// must be
const mergeWithStrategy = mergeWithCustomize({
    customizeObject: customizeObject({
        plugins: 'prepend',
    })
});
```

```js
plugins: [
    new CompressionPlugin({
        filename: '[path].gz[query]',
    }),

    new BrotliPlugin({
        asset: '[path].br[query]',
        test: /\.js$|\.css$|\.svg$|\.html$/,
        threshold: 10240,
        minRatio: 0.8
    })
]
// must be
plugins: [
    new CompressionPlugin({
        filename: '[path][base].br[query]',
        algorithm: 'brotliCompress',
        test: /\.(js|css|html|svg)$/,
        threshold: 10240,
        minRatio: 0.8,
    }),
]
```

In the `new TerserPlugin({ ... })` minimizer plugin, remove the `cache` property.

> 1. `frontend/libs/assets-configurator.js`:

Reviewer 2: ```suggestion
1. In `frontend/libs/assets-configurator.js`, make the following changes:
```

```

2. In `frontend/configs/development.js`, do the following:
   - Rename the `jsonpFunction` property to the `chunkLoadingGlobal` of the `output` object.
   - Rename the `vendors` property to the `defaultVendors` of the `cacheGroups` object.

3. In `frontend/configs/production.js`, make the following changes:

```js
const merge = require('webpack-merge');
// must be
const { mergeWithCustomize, customizeObject } = require('webpack-merge');
```

```js
const mergeWithStrategy = merge.smartStrategy({
    plugins: 'prepend',
});
// must be
const mergeWithStrategy = mergeWithCustomize({
    customizeObject: customizeObject({
        plugins: 'prepend',
    })
});
```

```js
plugins: [
    new CompressionPlugin({
        filename: '[path].gz[query]',
    }),

    new BrotliPlugin({
        asset: '[path].br[query]',
        test: /\.js$|\.css$|\.svg$|\.html$/,
        threshold: 10240,
        minRatio: 0.8
    })
]
// must be
plugins: [
    new CompressionPlugin({
        filename: '[path][base].br[query]',
        algorithm: 'brotliCompress',
        test: /\.(js|css|html|svg)$/,
        threshold: 10240,
        minRatio: 0.8,
    }),
]
```

> In the `new TerserPlugin({ ... })` minimizer plugin, remove the `cache` property.

Reviewer 2: @danilmoro3 should it be step 4? Or is it a sub-step of step 3?

Author: It is related to step 3

Author: It's not sub-item, please leave it in the same level with code samples

   - Rename the `vendors` property to the `defaultVendors` of the `cacheGroups` object.

3. `frontend/configs/production.js`:

   ```js
   const merge = require('webpack-merge');
   // must be
   const { mergeWithCustomize, customizeObject } = require('webpack-merge');
   ```

   ```js
   const mergeWithStrategy = merge.smartStrategy({
       plugins: 'prepend',
   });
   // must be
   const mergeWithStrategy = mergeWithCustomize({
       customizeObject: customizeObject({
           plugins: 'prepend',
       })
   });
   ```

   ```js
   plugins: [
       new CompressionPlugin({
           filename: '[path].gz[query]',
       }),
   
       new BrotliPlugin({
           asset: '[path].br[query]',
           test: /\.js$|\.css$|\.svg$|\.html$/,
           threshold: 10240,
           minRatio: 0.8
       })
   ]
   // must be
   plugins: [
       new CompressionPlugin({
           filename: '[path][base].br[query]',
           algorithm: 'brotliCompress',
           test: /\.(js|css|html|svg)$/,
           threshold: 10240,
           minRatio: 0.8,
       }),
   ]
   ```

   Remove the `cache` property in the `new TerserPlugin({ ... })` minimizer plugin.

> 

Reviewer: @VadymSachenko Removing indents impairs readability.

Reviewer 2: @FRITS-Kh from the perspective of UI, it does not have any impact. The content is displayed identically (with and without indents)

Reviewer: We work in the `Markdown` format and that makes sense for the format.
You can check it for example in GitHub or IDE preview.
Maybe the rule should be adjusted?

Reviewer 2: @FRITS-Kh, we must be consistent. Unfortunately, we can't do this. If we adjust this one, we must redo all the rest (we talk about hundreds of documents) but the reason for doing so must be stronger than just making the documents a bit prettier on GitHub. Also, idents might cause issues in Algolia, that's another reason we avoid them.

Reviewer 2: @FRITS-Kh Though we can apply indents for sub-steps because, in this case, they are displayed differently (right under the sub-step).
